<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto City</font></h1>
<h2 align=center><font size = 5>Part 3: Explore and cluster the neighborhoods in Toronto </font></h2>

## Introduction

In this lab, you will learn how to convert addresses into their equivalent latitude and longitude values. Also, you will use the Foursquare API to explore neighborhoods in Toronto City. You will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. You will use the *k*-means clustering algorithm to complete this task. Finally, you will use the Folium library to visualize the neighborhoods in Toronto City and their emerging clusters.

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [24]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')


Libraries imported.


### Read CSV genersted in part 2

In [71]:
neighborhoods = pd.read_csv('Toronto_Neighborhoods_with_coordinates.csv')
print('Data shape=' ,neighborhoods.shape)
neighborhoods.head()

Data shape= (103, 5)


Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

#### Use geopy library to get the latitude and longitude values of Toronto City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [26]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [27]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [28]:
CLIENT_ID = '5AHI1B14PNWWA3ZKPOQPLKSXRZ0P3OXLSUU1EMYDLVWWH1VY' # your Foursquare ID
CLIENT_SECRET = 'WSPO43X5JSJXEKH2F2X3MOUPFUQCBYLSY0WQYZL3FADJJAOR' # your Foursquare Secret
VERSION = '20180605'


print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5AHI1B14PNWWA3ZKPOQPLKSXRZ0P3OXLSUU1EMYDLVWWH1VY
CLIENT_SECRET:WSPO43X5JSJXEKH2F2X3MOUPFUQCBYLSY0WQYZL3FADJJAOR


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [99]:

neighborhoods.loc[0, 'Neighborhood']

'Parkwoods'

Get the neighborhood's latitude and longitude values.

In [100]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of "{}" are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of "Parkwoods" are 43.7532586, -79.3296565.


#### Now, let's get the top 100 venues that are in Parkwoods within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [101]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=5AHI1B14PNWWA3ZKPOQPLKSXRZ0P3OXLSUU1EMYDLVWWH1VY&client_secret=WSPO43X5JSJXEKH2F2X3MOUPFUQCBYLSY0WQYZL3FADJJAOR&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

Send the GET request and examine the resutls

In [102]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e6a5ae2edbcad001b9d74e4'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [33]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [34]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


D:\programs\Anaconda3\envs\tensorflow\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

And how many venues were returned by Foursquare?

In [35]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now Run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [37]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

toronto_venues.head()

Parkwoods
Victoria Village
Harbourfront
Lawrence Heights,Lawrence Manor
Queen's Park
Islington Avenue
Rouge,Malvern
Don Mills North
Woodbine Gardens,Parkview Hill
Ryerson,Garden District
Glencairn
Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park
Highland Creek,Rouge Hill,Port Union
Flemingdon Park,Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe
Guildwood,Morningside,West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,Downsview North,Wilson Heights
Thorncliffe Park
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Scarborough Village
Fairview,Henry Farm,Oriole
Northwood Park,York University
East Toronto
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
East Birchmount Park,Ionview,Kennedy Park
Bayview Village
CFB Toronto,Downsview East
The Danforth West,Riverdale
Design Exchange,Toro

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

#### Let's check the size of the resulting dataframe

In [103]:
print(toronto_venues.shape)


(2233, 7)


Let's check how many venues were returned for each neighborhood

In [49]:
toronto_venues.to_csv('toronto_venues.csv',index=False)

In [50]:
toronto_venues.groupby('Neighborhood')['Venue'].count().to_frame().head()

Venue
Neighborhood                                             
Adelaide,King,Richmond                                100
Agincourt                                               4
Agincourt North,L'Amoreaux East,Milliken,Steele...      2
Albion Gardens,Beaumond Heights,Humbergate,Jame...      9
Alderwood,Long Branch                                  10

#### Let's find out how many unique categories can be curated from all the returned venues

In [51]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [43]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()


Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                    0             0         0            0   
1                    0             0         0            0   
2                    0             0         0            0   
3                    0             0         0            0   
4                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0             0              0         0            0         0   
1 

In [64]:
# add neighborhood column back to dataframe      
toronto_onehot['Neighborhood'] =  toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns=list( toronto_onehot.columns)
fixed_columns.remove('Neighborhood')
fixed_columns =  ['Neighborhood'] + fixed_columns

toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0         Parkwoods         0          0            0              0   
1         Parkwoods         0          0            0              0   
2  Victoria Village         0          0            0              0   
3  Victoria Village         0          0            0              0   
4  Victoria Village         0          0            0              0   

   Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Te

And let's examine the new dataframe size.

In [66]:
toronto_onehot.shape

(2233, 267)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [68]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Wine Bar  Wine Shop  \
0                             Adelaide,King,Richmond      0.01        0.0   
1                                          Agincourt      0.00        0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...      0.00        0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...      0.00        0.0   
4                              Alderwood,Long Branch      0.00        0.0   

   Wings Joint  Women's Store  Yoga Studio  Accessories Store  \
0          0.0           0.01          0.0                0.0   
1          0.0           0.00          0.0                0.0   
2          0.0           0.00          0.0                0.0   
3          0.0           0.00          0.0                0.0   
4          0.0           0.00          0.0                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.02   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0          0.0                  0.0              0.02   
1           0.0       0.0          0.0                  0.0              0.00   
2           0.0       0.0          0.0                  0.0              0.00   
3           0.0       0.0          0.0                  0.0              0.00   
4           0.0       0.0          0.0                  0.0              0.00   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0            0.0        0.0         0.0         0.0   
1                 0.0            0.0        0.0         0.0         0.0   
2                 0.0            0.0        0.0         0.0         0.0   
3                 0.0            0.0        0.0         0.0         0.0   
4                 0.1            0.0        0.0         0.0         0.0   

   Bakery  Bank   Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0    0.02   0.0  0.03             0.0               0.0               0.0   
1    0.00   0.0  0.00             0.0               0.0               0.0   
2    0.00   0.0  0.00             0.0               0.0               0.0   
3    0.00   0.0  0.00             0.0               0.0               0.0   
4    0.00   0.0  0.00             0.0               0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0    0.000000   
1                 0.0    0.0              0.0       0.0    0.000000   
2                 0.0    0.0              0.0       0.0    0.000000   
3                 0.0    0.0              0.0       0.0    0.111111   
4                 0.0    0.0              0.0       0.0    0.000000   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0        0.0     0.0            0.0       0.02       0.0   
1                 0.0        0.0     0.0            0.0       0.00       0.0   
2                 0.0        0.0     0.0            0.0       0.00       0.0   
3                 0.0        0.0     0.0            0.0       0.00       0.0   
4                 0.0        0.0     0.0            0.0       0.00       0.0   

   Brazilia

#### Let's confirm the new size

In [69]:
toronto_grouped.shape

(100, 267)

### Notes: as shown Neighborhoods are 100 instead of 103 as there are three  Neighborhoods has no Venues

In [104]:
#Three Neighborhood has no Venues
a = toronto_venues['Neighborhood'].unique()
b = neighborhoods['Neighborhood'].unique()
neighborhood_has_no_venues= set(b) - set(a)
neighborhood_has_no_venues

{'Islington Avenue', 'Newtonbrook,Willowdale', 'Upper Rouge'}

#### Let's print each neighborhood along with the top 5 most common venues

In [105]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1       Restaurant  0.05
2             Café  0.04
3  Thai Restaurant  0.04
4              Bar  0.03


----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1             Clothing Store  0.25
2                     Lounge  0.25
3             Breakfast Spot  0.25
4             Massage Studio  0.00


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
           venue  freq
0     Playground   0.5
1           Park   0.5
2       Wine Bar   0.0
3  Metro Station   0.0
4    Men's Store   0.0


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                 venue  freq
0        Grocery Store  0.22
1          Pizza Place  0.22
2           Beer Store  0.11
3       Sandwich Place  0.11
4  Fried Chicken Joint  0.11


----Alderwood,Long Branch----
                venue  freq
0         Pizza Place   0.2
1           

               venue  freq
0        Coffee Shop  0.25
1   Department Store  0.25
2  Convenience Store  0.25
3     Discount Store  0.25
4     Medical Center  0.00


----East Toronto----
               venue  freq
0               Park  0.50
1      Metro Station  0.25
2  Convenience Store  0.25
3           Wine Bar  0.00
4     Massage Studio  0.00


----Emery,Humberlea----
                venue  freq
0      Baseball Field   1.0
1            Wine Bar   0.0
2              Market   0.0
3  Mexican Restaurant   0.0
4       Metro Station   0.0


----Fairview,Henry Farm,Oriole----
                  venue  freq
0        Clothing Store  0.13
1           Coffee Shop  0.07
2  Fast Food Restaurant  0.06
3         Women's Store  0.04
4   Japanese Restaurant  0.04


----First Canadian Place,Underground city----
         venue  freq
0  Coffee Shop  0.12
1         Café  0.07
2   Restaurant  0.06
3    Gastropub  0.03
4          Gym  0.03


----Flemingdon Park,Don Mills South----
              venue  freq


            venue  freq
0  Sandwich Place  0.14
1            Café  0.14
2     Coffee Shop  0.09
3       BBQ Joint  0.05
4  Cosmetics Shop  0.05


----The Beaches----
                venue  freq
0                 Pub  0.25
1   Health Food Store  0.25
2               Trail  0.25
3      Massage Studio  0.00
4  Mexican Restaurant  0.00


----The Beaches West,India Bazaar----
                  venue  freq
0        Sandwich Place  0.06
1      Sushi Restaurant  0.06
2    Italian Restaurant  0.06
3         Burrito Place  0.06
4  Fast Food Restaurant  0.06


----The Danforth West,Riverdale----
                    venue  freq
0        Greek Restaurant  0.20
1             Coffee Shop  0.10
2      Italian Restaurant  0.07
3  Furniture / Home Store  0.05
4          Ice Cream Shop  0.05


----The Junction North,Runnymede----
                       venue  freq
0                    Brewery  0.25
1       Caribbean Restaurant  0.25
2              Grocery Store  0.25
3                   Bus Line  0.25
4 

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [106]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [107]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt        Clothing Store   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                  Park   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Restaurant                  Café       Thai Restaurant   
1  Latin American Restaurant                Lounge        Breakfast Spot   
2                 Playground       Warehouse Store      Department Store   
3                Pizza Place            Beer Store  Fast Food Restaurant   
4                       Pool                   Gym          Skating Rink   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar            Steakhouse      Sushi Restaurant   
1          Cupcake Shop      Doner Restaurant               Dog Run   
2      Cuban Restaurant          Cupcake Shop           Curling Ice   
3   Fried Chicken Joint              Pharmacy        Sandwich Place   
4                   Pub    Athletics & Sports              Pharmacy   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel                   Gym              Gastropub  
1   Distribution Center        Discount Store                  Diner  
2          Dance Studio         Deli / Bodega           Dessert Shop  
3                 Diner        Discount Store     Dim Sum Restaurant  
4        Sandwich Place           Coffee Shop        College Stadium

In [108]:
toronto_grouped.head()

Neighborhood  Wine Bar  Wine Shop  \
0                             Adelaide,King,Richmond      0.01        0.0   
1                                          Agincourt      0.00        0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...      0.00        0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...      0.00        0.0   
4                              Alderwood,Long Branch      0.00        0.0   

   Wings Joint  Women's Store  Yoga Studio  Accessories Store  \
0          0.0           0.01          0.0                0.0   
1          0.0           0.00          0.0                0.0   
2          0.0           0.00          0.0                0.0   
3          0.0           0.00          0.0                0.0   
4          0.0           0.00          0.0                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.02   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0          0.0                  0.0              0.02   
1           0.0       0.0          0.0                  0.0              0.00   
2           0.0       0.0          0.0                  0.0              0.00   
3           0.0       0.0          0.0                  0.0              0.00   
4           0.0       0.0          0.0                  0.0              0.00   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0            0.0        0.0         0.0         0.0   
1                 0.0            0.0        0.0         0.0         0.0   
2                 0.0            0.0        0.0         0.0         0.0   
3                 0.0            0.0        0.0         0.0         0.0   
4                 0.1            0.0        0.0         0.0         0.0   

   Bakery  Bank   Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0    0.02   0.0  0.03             0.0               0.0               0.0   
1    0.00   0.0  0.00             0.0               0.0               0.0   
2    0.00   0.0  0.00             0.0               0.0               0.0   
3    0.00   0.0  0.00             0.0               0.0               0.0   
4    0.00   0.0  0.00             0.0               0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0    0.000000   
1                 0.0    0.0              0.0       0.0    0.000000   
2                 0.0    0.0              0.0       0.0    0.000000   
3                 0.0    0.0              0.0       0.0    0.111111   
4                 0.0    0.0              0.0       0.0    0.000000   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0        0.0     0.0            0.0       0.02       0.0   
1                 0.0        0.0     0.0            0.0       0.00       0.0   
2                 0.0        0.0     0.0            0.0       0.00       0.0   
3                 0.0        0.0     0.0            0.0       0.00       0.0   
4                 0.0        0.0     0.0            0.0       0.00       0.0   

   Brazilia

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [109]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 4, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [112]:
neighborhoods.head()

Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [113]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt        Clothing Store   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                  Park   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Restaurant                  Café       Thai Restaurant   
1  Latin American Restaurant                Lounge        Breakfast Spot   
2                 Playground       Warehouse Store      Department Store   
3                Pizza Place            Beer Store  Fast Food Restaurant   
4                       Pool                   Gym          Skating Rink   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar            Steakhouse      Sushi Restaurant   
1          Cupcake Shop      Doner Restaurant               Dog Run   
2      Cuban Restaurant          Cupcake Shop           Curling Ice   
3   Fried Chicken Joint              Pharmacy        Sandwich Place   
4                   Pub    Athletics & Sports              Pharmacy   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel                   Gym              Gastropub  
1   Distribution Center        Discount Store                  Diner  
2          Dance Studio         Deli / Bodega           Dessert Shop  
3                 Diner        Discount Store     Dim Sum Restaurant  
4        Sandwich Place           Coffee Shop        College Stadium

In [121]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted['Cluster Labels']=kmeans.labels_
toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(10) # check the last columns!

Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   
5      M9A         Etobicoke                 Islington Avenue  43.667856   
6      M1B       Scarborough                    Rouge,Malvern  43.806686   
7      M3B        North York                  Don Mills North  43.745906   
8      M4B         East York   Woodbine Gardens,Parkview Hill  43.706397   
9      M5B  Downtown Toronto          Ryerson,Garden District  43.657162   

   Longitude 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656     Food & Drink Shop                    Park   
1 -79.315572          Hockey Arena            Intersection   
2 -79.360636           Coffee Shop                     Pub   
3 -79.464763        Clothing Store  Furniture / Home Store   
4 -79.389494           Coffee Shop                    Park   
5 -79.532242                   NaN                     NaN   
6 -79.194353            Print Shop    Fast Food Restaurant   
7 -79.352188        Baseball Field     Japanese Restaurant   
8 -79.309937           Pizza Place      Athletics & Sports   
9 -79.378937           Coffee Shop          Clothing Store   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0               Dessert Shop      Cuban Restaurant          Cupcake Shop   
1      Portuguese Restaurant           Pizza Place           Coffee Shop   
2                       Café                Bakery                  Park   
3                   Boutique           Event Space    Miscellaneous Shop   
4                   Beer Bar          Burger Joint                  Café   
5                        NaN                   NaN                   NaN   
6           Cuban Restaurant          Cupcake Shop           Curling Ice   
7           Basketball Court                  Café  Gym / Fitness Center   
8                   Bus Line             Gastropub        Breakfast Spot   
9  Middle Eastern Restaurant                  Café   Japanese Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Curling Ice          Dance Studio         Deli / Bodega   
1               Dog Run   Distribution Center        Discount Store   
2        Breakfast Spot    Mexican Restaurant               Theater   
3           Coffee Shop          Carpet Store             Gift Shop   
4  Fast Food Restaurant              Creperie                   Bar   
5                   NaN                   NaN                   NaN   
6          Dance Studio         Deli / Bodega      Department Store   
7  Caribbean Restaurant                 Diner         Deli / Bodega   
8                  Bank          Intersection  Gym / Fitness Center   
9       Bubble Tea Shop        Lingerie Store      Ramen Restaurant   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
0       Department Store     Dim Sum Restaurant             1.0  
1       Doner Restaurant                  Diner             0.0  
2             Beer Store            Event Space             0.0  
3  Vietnamese Restaurant      Accessories Store             0.0  
4     Seafood Restaurant          Burrito Place             0.0  
5                    NaN                    NaN             NaN  
6           Dessert Shop              Drugstore             0.0  
7       Department Store           Dessert Shop             0.0  
8   Fast Food Restaurant               Pharmacy             0.0  
9     Italian Restaurant         Cosmetics Shop             0.0

In [124]:
toronto_merged['Cluster Labels'].isnull().sum()
print(toronto_merged.shape)
# dropping rows with  null values 
toronto_merged.dropna(axis = 0, inplace = True) 
print(toronto_merged.shape)


(103, 16)
(100, 16)


Finally, let's visualize the resulting clusters

In [125]:
x = np.arange(kclusters)
print(x)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
rainbow

[0 1 2 3 4]


['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']

In [127]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    #print(cluster)
    cluster = int(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [128]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0 , toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue      2nd Most Common Venue  \
1          North York          Hockey Arena               Intersection   
2    Downtown Toronto           Coffee Shop                        Pub   
3          North York        Clothing Store     Furniture / Home Store   
4    Downtown Toronto           Coffee Shop                       Park   
6         Scarborough            Print Shop       Fast Food Restaurant   
7          North York        Baseball Field        Japanese Restaurant   
8           East York           Pizza Place         Athletics & Sports   
9    Downtown Toronto           Coffee Shop             Clothing Store   
10         North York           Pizza Place                        Pub   
12        Scarborough        History Museum                        Bar   
13         North York           Coffee Shop                 Restaurant   
14          East York          Skating Rink             Cosmetics Shop   
15   Downtown Toronto           Coffee Shop                       Café   
17          Etobicoke        Cosmetics Shop                       Park   
18        Scarborough                   Spa             Breakfast Spot   
19       East Toronto                   Pub          Health Food Store   
20   Downtown Toronto           Coffee Shop               Cocktail Bar   
22        Scarborough           Coffee Shop          Korean Restaurant   
23          East York   Sporting Goods Shop                Coffee Shop   
24   Downtown Toronto           Coffee Shop         Italian Restaurant   
25   Downtown Toronto         Grocery Store                       Café   
26        Scarborough       Thai Restaurant                     Lounge   
27         North York  Fast Food Restaurant                    Dog Run   
28         North York           Coffee Shop                  Gift Shop   
29          East York     Indian Restaurant            Warehouse Store   
30   Downtown Toronto           Coffee Shop                 Restaurant   
31       West Toronto                Bakery                   Pharmacy   
33         North York        Clothing Store                Coffee Shop   
34         North York           Coffee Shop         Falafel Restaurant   
36   Downtown Toronto           Coffee Shop                   Aquarium   
37       West Toronto                   Bar                Coffee Shop   
38        Scarborough        Discount Store                Coffee Shop   
39         North York                  Bank        Japanese Restaurant   
41       East Toronto      Greek Restaurant                Coffee Shop   
42   Downtown Toronto           Coffee Shop                       Café   
43       West Toronto                  Café                Coffee Shop   
44        Scarborough              Bus Line                     Bakery   
45         North York             Cafeteria            Warehouse Store   
46         North York         Grocery Store                       Bank   
47       East Toronto    Italian Restaurant                       Park   
48   Downtown Toronto           Coffee Shop                 Restaurant   
50         North York           Pizza Place        Empanada Restaurant   
51        Scarborough   American Restaurant                      Motel   
54       East Toronto                  Café                Coffee Shop   
55         North York           Coffee Shop             Sandwich Place   
56               York            Restaurant             Sandwich Place   
58        Scarborough                  Café      General Entertainment   
59         North York      Ramen Restaurant             Sandwich Place   
60         North York          Liquor Store              Grocery Store   
62    Central Toronto                Garden         Dim Sum Restaurant   
63               York  Caribbean Restaurant              Grocery Store   
64               York     Convenience Store            Warehouse Store   
65        Scarborough     Indian Restaurant      Vietnamese Restaurant   
67    Central Toro

#### Cluster 2

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 1 ] ) 

#### Cluster 3

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 2 ] ) 

#### Cluster 4

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 3 ] ) 

#### Cluster 5

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 4 ] ) 

## Name each cluster based on most common Venues in this cluster

In [137]:
#this function will count Most Common Venues 1--> 10 
def cluster_name(data):
    ven = []
    for i in range(4,14): # from 1st Most Common Venue to 10th  Most Common Venue  
        ven= ven+ list(data[data.columns[i]]) # add all Venues in one list  
    ven = pd.DataFrame({'Venue':ven})
    return ven.Venue.value_counts().head() # return most 5 Venues

In [138]:
#dislay 
for i in range(0,5):
    # return most common Venue 
    print('Cluster_No:',i)
    print(cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == i ] ) )
    print()

Cluster_No: 0
Coffee Shop      43
Café             30
Deli / Bodega    21
Dance Studio     21
Restaurant       20
Name: Venue, dtype: int64

Cluster_No: 1
Park             11
Dance Studio     10
Cupcake Shop      9
Curling Ice       9
Deli / Bodega     8
Name: Venue, dtype: int64

Cluster_No: 2
Golf Course         1
Department Store    1
Dance Studio        1
Warehouse Store     1
Dessert Shop        1
Name: Venue, dtype: int64

Cluster_No: 3
Department Store    3
Dance Studio        3
Dessert Shop        3
Curling Ice         3
Baseball Field      3
Name: Venue, dtype: int64

Cluster_No: 4
Department Store    3
Dance Studio        3
Playground          3
Cupcake Shop        3
Curling Ice         3
Name: Venue, dtype: int64



**Conclusion**:
we can name cluster from most two/three Venues as followig:
- Cluster_0: Coffee shops area
- Cluster_1: Park  and Dance area 
- Cluster_2: Miscellaneous (as all Venues not repated  )  
- Cluster_3: Department Store ,Dance Studio ,Dessert Shop  
- Cluster_4: Department Store ,Dance Studio ,Playground           